src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+      (is-eq owner contract-caller)
+      (err ERR_NAME_OPERATION_UNAUTHORIZED))
+    ;; The name must not be in the renewal grace period
+    (asserts!
+      (is-eq (unwrap! (is-name-in-grace-period? namespace name) (err ERR_PANIC)) false)
+      (err ERR_NAME_GRACE_PERIOD))
+    ;; The name must not be expired 
+    (asserts!
+      (is-eq (unwrap! (is-name-lease-expired? namespace name) (err ERR_PANIC)) false)
+      (err ERR_NAME_EXPIRED))
+    ;; The name must not be revoked
+    (asserts!
+      (is-none (get revoked-at name-props))
+      (err ERR_NAME_REVOKED))
+    ;; Update the zonefile
+    (update-zonefile-and-props
+      namespace 
+      name  
+      (get registered-at name-props)
+      (get imported-at name-props)
+      none
+      zonefile-hash)
+    (ok true)))
+
+;; NAME_TRANSFER
+;; A NAME_TRANSFER transaction changes the name's public key hash. You would send one of these transactions if you wanted to:
+;; - Change your private key
+;; - Send the name to someone else
+;; When transferring a name, you have the option to also clear the name's zone file hash (i.e. set it to null). 
+;; This is useful for when you send the name to someone else, so the recipient's name does not resolve to your zone file.
+(define-public (name-transfer (namespace (buff 19))
+                              (name (buff 16))
+                              (new-owner principal)
+                              (zonefile-hash (optional (buff 20))))
+  (let (
+    (current-owned-name (map-get? owner-name ((owner new-owner))))
+    (namespace-props (unwrap!
+      (map-get? namespaces ((namespace namespace)))
+      (err ERR_NAMESPACE_NOT_FOUND))))
+    (let (
+      (owner (unwrap!
+        (nft-get-owner? names (tuple (name name) (namespace namespace)))
+        (err ERR_NAME_NOT_FOUND))) ;; The name must exist
+      (name-props (unwrap!
+        (map-get? name-properties ((name name) (namespace namespace)))
+        (err ERR_NAME_NOT_FOUND))) ;; The name must exist
+      (can-new-owner-get-name (if (is-none current-owned-name)
+                                  true
+                                  (unwrap! 
+                                    (is-name-lease-expired?
+                                      (unwrap! (get namespace current-owned-name) (err ERR_PANIC))
+                                      (unwrap! (get name current-owned-name) (err ERR_PANIC)))
+                                    (err ERR_PANIC)))))
+      ;; The namespace must be launched
+      (asserts!
+        (is-some (get launched-at namespace-props))
+        (err ERR_NAMESPACE_NOT_LAUNCHED))
+      ;; The sender must match the name's current owner
+      (asserts!
+        (is-eq owner contract-caller)
+        (err ERR_NAME_OPERATION_UNAUTHORIZED))
+      ;; The name must have been registered (vs imported)
+      (unwrap!
+        (get registered-at name-props)
+        (err ERR_NAME_OPERATION_UNAUTHORIZED))
+      ;; The name must not be in the renewal grace period
```

Comment:
I think you may be able to combine the checks that a name is not in expired, not revoked, and not in the renewal grace period into a single `define-read-only` function.  This would be a nice API endpoint to have, and would also remove some code duplication here.

Reply:
There's already the `name-resolve` read-only endpoint, achieving this.
Code de-duplication can also be addressed by addressing this feedback from @kantai:

> We probably don't need update/transfer to be separate operations anymore, as there's not much savings between the two, and we probably don't need the renew function to also optionally be able to perform a transfer.

Thoughts?

Reply:
I agree with @kantai.  We could merge the update and transfer methods into a single method, and have helper wrappers that do only an update or only a transfer.  This would achieve the same thing.